File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/stack/upscale/StackUpscaleService.java
package com.sequenceiq.cloudbreak.core.flow2.stack.upscale;

import static com.sequenceiq.cloudbreak.api.endpoint.v4.common.Status.AVAILABLE;
import static com.sequenceiq.cloudbreak.api.endpoint.v4.common.Status.UPDATE_FAILED;
import static com.sequenceiq.cloudbreak.api.endpoint.v4.common.Status.UPDATE_IN_PROGRESS;
import static com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.base.InstanceStatus.CREATED;
import static com.sequenceiq.cloudbreak.event.ResourceEvent.STACK_ADDING_INSTANCES;
import static com.sequenceiq.cloudbreak.event.ResourceEvent.STACK_INFRASTRUCTURE_UPDATE_FAILED;
import static com.sequenceiq.cloudbreak.event.ResourceEvent.STACK_METADATA_EXTEND_WITH_COUNT;
import static com.sequenceiq.cloudbreak.event.ResourceEvent.STACK_REPAIR_FAILED;
import static com.sequenceiq.cloudbreak.event.ResourceEvent.STACK_UPSCALE_IMAGE_FALLBACK;
import static com.sequenceiq.cloudbreak.event.ResourceEvent.STACK_UPSCALE_QUOTA_ISSUE;
import static com.sequenceiq.cloudbreak.event.ResourceEvent.STACK_VERTICALSCALE_ISSUE;
import static java.lang.String.format;

import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

import jakarta.inject.Inject;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import com.sequenceiq.cloudbreak.api.endpoint.v4.common.DetailedStackStatus;
import com.sequenceiq.cloudbreak.cloud.CloudConnector;
import com.sequenceiq.cloudbreak.cloud.UpdateType;
import com.sequenceiq.cloudbreak.cloud.context.AuthenticatedContext;
import com.sequenceiq.cloudbreak.cloud.event.instance.CollectMetadataResult;
import com.sequenceiq.cloudbreak.cloud.exception.CloudConnectorException;
import com.sequenceiq.cloudbreak.cloud.exception.QuotaExceededException;
import com.sequenceiq.cloudbreak.cloud.model.CloudInstance;
import com.sequenceiq.cloudbreak.cloud.model.CloudResourceStatus;
import com.sequenceiq.cloudbreak.cloud.model.CloudStack;
import com.sequenceiq.cloudbreak.cloud.model.CloudVmMetaDataStatus;
import com.sequenceiq.cloudbreak.cloud.model.Group;
import com.sequenceiq.cloudbreak.cloud.model.InstanceStoreMetadata;
import com.sequenceiq.cloudbreak.cloud.model.ResourceStatus;
import com.sequenceiq.cloudbreak.common.exception.CloudbreakServiceException;
import com.sequenceiq.cloudbreak.common.service.TransactionService;
import com.sequenceiq.cloudbreak.common.service.TransactionService.TransactionExecutionException;
import com.sequenceiq.cloudbreak.core.flow2.stack.CloudbreakFlowMessageService;
import com.sequenceiq.cloudbreak.core.flow2.stack.downscale.StackScalingFlowContext;
import com.sequenceiq.cloudbreak.domain.stack.instance.InstanceMetaData;
import com.sequenceiq.cloudbreak.dto.StackDto;
import com.sequenceiq.cloudbreak.event.ResourceEvent;
import com.sequenceiq.cloudbreak.reactor.api.event.resource.CoreVerticalScaleRequest;
import com.sequenceiq.cloudbreak.reactor.api.event.resource.CoreVerticalScaleResult;
import com.sequenceiq.cloudbreak.reactor.api.event.stack.UpscaleStackRequest;
import com.sequenceiq.cloudbreak.reactor.api.event.stack.UpscaleStackResult;
import com.sequenceiq.cloudbreak.service.CloudbreakException;
import com.sequenceiq.cloudbreak.service.OperationException;
import com.sequenceiq.cloudbreak.service.StackUpdater;
import com.sequenceiq.cloudbreak.service.cluster.ClusterService;
import com.sequenceiq.cloudbreak.service.stack.InstanceMetaDataService;
import com.sequenceiq.cloudbreak.service.stack.flow.InstanceMetadataInstanceIdUpdater;
import com.sequenceiq.cloudbreak.service.stack.flow.MetadataSetupService;
import com.sequenceiq.cloudbreak.service.stack.flow.TlsSetupService;
import com.sequenceiq.cloudbreak.view.InstanceMetadataView;
import com.sequenceiq.cloudbreak.view.StackView;
import com.sequenceiq.common.api.adjustment.AdjustmentTypeWithThreshold;
import com.sequenceiq.common.api.type.CommonResourceType;

@Service
public class StackUpscaleService {

    private static final Logger LOGGER = LoggerFactory.getLogger(StackUpscaleService.class);

    private static final double ONE_HUNDRED = 100;

    @Inject
    private StackUpdater stackUpdater;

    @Inject
    private CloudbreakFlowMessageService flowMessageService;

    @Inject
    private MetadataSetupService metadataSetupService;

    @Inject
    private ClusterService clusterService;

    @Inject
    private TlsSetupService tlsSetupService;

    @Inject
    private TransactionService transactionService;

    @Inject
    private StackScalabilityCondition stackScalabilityCondition;

    @Inject
    private InstanceMetaDataService instanceMetaDataService;

    @Inject
    private InstanceMetadataInstanceIdUpdater instanceMetadataInstanceIdUpdater;

    public void startAddInstances(StackView stack, Map<String, Integer> hostGroupWithAdjustment) {
        Integer scalingAdjustment = hostGroupWithAdjustment.values().stream().reduce(0, Integer::sum);
        String statusReason = format("Adding %s new instance(s) to instance group %s", scalingAdjustment,
                String.join(", ", hostGroupWithAdjustment.keySet()));
        stackUpdater.updateStackStatus(stack.getId(), DetailedStackStatus.ADDING_NEW_INSTANCES, statusReason);
    }

    public void addInstanceFireEventAndLog(StackView stack, Map<String, Integer> hostGroupWithAdjustment,
            AdjustmentTypeWithThreshold adjustmentTypeWithThreshold) {
        Integer scalingAdjustment = hostGroupWithAdjustment.values().stream().reduce(0, Integer::sum);
        flowMessageService.fireEventAndLog(stack.getId(), UPDATE_IN_PROGRESS.name(), STACK_ADDING_INSTANCES, String.valueOf(scalingAdjustment),
                String.join(", ", hostGroupWithAdjustment.keySet()), String.valueOf(adjustmentTypeWithThreshold.getAdjustmentType()),
                String.valueOf(adjustmentTypeWithThreshold.getThreshold()));
    }

    public void finishAddInstances(StackScalingFlowContext context, UpscaleStackResult payload) {
        LOGGER.debug("Upscale stack result: {}", payload);
        List<CloudResourceStatus> results = payload.getResults();
        validateResourceResults(context, payload.getErrorDetails(), results);
        Set<Long> successfulPrivateIds = results.stream()
                .filter(cloudResourceStatus -> !ResourceStatus.FAILED.equals(cloudResourceStatus.getStatus()))
                .map(CloudResourceStatus::getPrivateId).filter(Objects::nonNull).collect(Collectors.toSet());
        Set<String> instanceGroups = context.getHostGroupWithAdjustment().keySet();
        metadataSetupService.cleanupRequestedInstancesIfNotInList(context.getStackId(), instanceGroups, successfulPrivateIds);
        if (successfulPrivateIds.isEmpty()) {
            metadataSetupService.cleanupRequestedInstancesWithoutFQDN(context.getStackId(), instanceGroups);
            throw new OperationException("Failed to upscale the cluster since all create request failed. Resource set is empty");
        }
        LOGGER.debug("Adding new instances to the stack is DONE");
    }

    public void extendingMetadata(Long stackId) {
        stackUpdater.updateStackStatus(stackId, DetailedStackStatus.EXTENDING_METADATA);
    }

    void reRegisterWithClusterProxy(long stackId) {
        clusterService.updateClusterStatusByStackId(stackId, DetailedStackStatus.REGISTERING_TO_CLUSTER_PROXY,
                "Re-registering with Cluster Proxy service.");
        flowMessageService.fireEventAndLog(stackId, UPDATE_IN_PROGRESS.name(), ResourceEvent.CLUSTER_RE_REGISTER_WITH_CLUSTER_PROXY);
    }

    public Set<String> finishExtendMetadata(StackView stackView, Integer scalingAdjustment, CollectMetadataResult payload) throws TransactionExecutionException {
        return transactionService.required(() -> {
            List<CloudVmMetaDataStatus> coreInstanceMetaData = payload.getResults();
            metadataSetupService.saveInstanceMetaData(stackView, coreInstanceMetaData, CREATED);
            Set<String> upscaleCandidateAddresses = coreInstanceMetaData.stream().filter(im -> im.getMetaData().getPrivateIp() != null)
                    .map(im -> im.getMetaData().getPrivateIp()).collect(Collectors.toSet());
            int newInstances = upscaleCandidateAddresses.size();
            try {
                clusterService.updateClusterStatusByStackIdOutOfTransaction(stackView.getId(), DetailedStackStatus.EXTENDING_METADATA_FINISHED);
            } catch (TransactionExecutionException e) {
                throw e.getCause();
            }
            if (scalingAdjustment != newInstances) {
                flowMessageService.fireEventAndLog(stackView.getId(), AVAILABLE.name(), STACK_METADATA_EXTEND_WITH_COUNT, String.valueOf(newInstances),
                        String.valueOf(scalingAdjustment));
            }
            return upscaleCandidateAddresses;
        });
    }

    public void setupTls(StackDto stack) throws CloudbreakException {
        for (InstanceMetadataView gwInstance : stack.getAllAvailableGatewayInstances()) {
            if (!stack.getTunnel().useCcm() && CREATED.equals(gwInstance.getInstanceStatus())) {
                tlsSetupService.setupTls(stack, gwInstance);
            }
        }
    }

    public void bootstrappingNewNodes(StackView stack) {
        stackUpdater.updateStackStatus(stack.getId(), DetailedStackStatus.BOOTSTRAPPING_NEW_NODES);
    }

    public void extendingHostMetadata(StackView stack) {
        stackUpdater.updateStackStatus(stack.getId(), DetailedStackStatus.EXTENDING_HOST_METADATA);
    }

    public void finishExtendHostMetadata(StackView stack) {
        stackUpdater.updateStackStatus(stack.getId(), DetailedStackStatus.STACK_UPSCALE_COMPLETED, "Stack upscale has been finished successfully.");
    }

    public void handleStackUpscaleFailure(Boolean upscaleForRepair, Map<String, Set<String>> hostgroupWithHostnames, Exception exception, Long stackId,
            Map<String, Integer> hostGroupWithAdjustment) {
        LOGGER.info("Exception during the upscale of stack", exception);
        try {
            String errorReason = exception != null ? exception.getMessage() : "Error reason is unknown";
            if (!upscaleForRepair) {
                metadataSetupService.cleanupRequestedInstancesWithoutFQDN(stackId, hostGroupWithAdjustment.keySet());
                stackUpdater.updateStackStatus(stackId, DetailedStackStatus.UPSCALE_FAILED, "Stack update failed. " + errorReason);
                flowMessageService.fireEventAndLog(stackId, UPDATE_FAILED.name(), STACK_INFRASTRUCTURE_UPDATE_FAILED, errorReason);
            } else {
                metadataSetupService.handleRepairFail(stackId, hostgroupWithHostnames.values().stream()
                        .flatMap(Collection::stream).collect(Collectors.toSet()));
                stackUpdater.updateStackStatus(stackId, DetailedStackStatus.REPAIR_FAILED, "Stack repair failed. " + errorReason);
                flowMessageService.fireEventAndLog(stackId, UPDATE_FAILED.name(), STACK_REPAIR_FAILED, errorReason);
            }
        } catch (Exception e) {
            LOGGER.info("Exception during the handling of stack scaling failure: {}", e.getMessage());
        }
    }

    private void validateResourceResults(StackScalingFlowContext context, Exception exception, List<CloudResourceStatus> results) {
        if (exception != null) {
            LOGGER.info(format("Failed to upscale stack: %s", context.getCloudContext()), exception);
            throw new OperationException(exception);
        }
        List<CloudResourceStatus> templates = results.stream().filter(result -> CommonResourceType.TEMPLATE == result.getCloudResource().getType()
                .getCommonResourceType()).collect(Collectors.toList());
        if (!templates.isEmpty() && (templates.get(0).isFailed() || templates.get(0).isDeleted())) {
            throw new OperationException(format("Failed to upscale the stack for %s due to: %s",
                    context.getCloudContext(), templates.get(0).getStatusReason()));
        }
    }

    public int getInstanceCountToCreate(StackDto stack, String instanceGroupName, int adjustment, boolean repair) {
        Set<InstanceMetaData> unusedInstanceMetadata = instanceMetaDataService.unusedInstancesInInstanceGroupByName(stack.getId(), instanceGroupName);
        int reusableInstanceCount = repair ? 0 : unusedInstanceMetadata.size();
        return stackScalabilityCondition.isScalable(stack, instanceGroupName) ? adjustment - reusableInstanceCount : 0;
    }

    public List<CloudResourceStatus> upscale(AuthenticatedContext ac, UpscaleStackRequest<UpscaleStackResult> request, CloudConnector connector)
            throws QuotaExceededException, TransactionExecutionException {
        CloudStack cloudStack = request.getCloudStack();
        AdjustmentTypeWithThreshold adjustmentTypeWithThreshold = request.getAdjustmentWithThreshold();
        try {
            return upscaleAndSyncStatusesToInstanceMetadata(ac, request, connector, cloudStack, adjustmentTypeWithThreshold);
        } catch (QuotaExceededException quotaExceededException) {
            return handleQuotaExceptionAndRetryUpscale(request, connector, ac, cloudStack, adjustmentTypeWithThreshold, quotaExceededException);
        }
    }

    public List<CloudResourceStatus> verticalScale(AuthenticatedContext ac, CoreVerticalScaleRequest<CoreVerticalScaleResult> request,
            CloudConnector connector, String group) throws Exception {
        return verticalScale(ac, request, connector, group, UpdateType.VERTICAL_SCALE);
    }

    public List<CloudResourceStatus> verticalScaleWithoutInstances(AuthenticatedContext ac, CoreVerticalScaleRequest<CoreVerticalScaleResult> request,
            CloudConnector connector, String group) throws Exception {
        return verticalScale(ac, request, connector, group, UpdateType.VERTICAL_SCALE_WITHOUT_INSTANCES);
    }

    private List<CloudResourceStatus> upscaleAndSyncStatusesToInstanceMetadata(AuthenticatedContext ac, UpscaleStackRequest<UpscaleStackResult> request,
            CloudConnector connector, CloudStack cloudStack, AdjustmentTypeWithThreshold adjustmentTypeWithThreshold)
            throws QuotaExceededException, TransactionExecutionException {
        List<CloudResourceStatus> cloudResourceStatuses = connector.resources()
                .upscale(ac, cloudStack, request.getResourceList(), adjustmentTypeWithThreshold);
        instanceMetadataInstanceIdUpdater.updateWithInstanceIdAndStatus(ac, connector, cloudResourceStatuses);
        return cloudResourceStatuses;
    }

    private List<CloudResourceStatus> handleQuotaExceptionAndRetryUpscale(UpscaleStackRequest<UpscaleStackResult> request, CloudConnector connector,
            AuthenticatedContext ac, CloudStack cloudStack, AdjustmentTypeWithThreshold adjustmentTypeWithThreshold,
            QuotaExceededException quotaExceededException) throws QuotaExceededException, TransactionExecutionException {
        flowMessageService.fireEventAndLog(request.getResourceId(), UPDATE_IN_PROGRESS.name(), STACK_UPSCALE_QUOTA_ISSUE,
                quotaExceededException.getQuotaErrorMessage());
        List<Group> groups = cloudStack.getGroups();
        int removableNodeCount = getRemovableNodeCount(adjustmentTypeWithThreshold, quotaExceededException, groups);
        decreaseInstances(groups, removableNodeCount);
        return upscaleAndSyncStatusesToInstanceMetadata(ac, request, connector, cloudStack, adjustmentTypeWithThreshold);
    }

    private List<CloudResourceStatus> verticalScale(AuthenticatedContext ac, CoreVerticalScaleRequest<CoreVerticalScaleResult> request,
            CloudConnector connector, String group, UpdateType updateType) throws Exception {
        CloudStack cloudStack = request.getCloudStack();
        try {
            return connector.resources().update(ac, cloudStack, request.getResourceList(), updateType,
                    Optional.ofNullable(group));
        } catch (Exception e) {
            return handleExceptionAndRetryUpdate(request, connector, ac, cloudStack, e, updateType, group);
        }
    }

    private List<CloudResourceStatus> handleExceptionAndRetryUpdate(CoreVerticalScaleRequest<CoreVerticalScaleResult> request,
            CloudConnector connector, AuthenticatedContext ac, CloudStack cloudStack, Exception e, UpdateType type, String group) throws Exception {
        LOGGER.error("Failed to vertically scale the stack", e);
        flowMessageService.fireEventAndLog(request.getResourceId(), UPDATE_IN_PROGRESS.name(), STACK_VERTICALSCALE_ISSUE,
                e.getMessage());
        return connector.resources().update(ac, cloudStack, request.getResourceList(), type, Optional.ofNullable(group));
    }

    private int getRemovableNodeCount(AdjustmentTypeWithThreshold adjustmentTypeWithThreshold, QuotaExceededException quotaExceededException,
            List<Group> groups) {
        int additionalRequired = quotaExceededException.getAdditionalRequired();
        int originalRequestedNodeCount = groups.stream().flatMap(group -> group.getInstances().stream())
                .filter(cloudInstance -> cloudInstance.getInstanceId() == null)
                .collect(Collectors.toList())
                .size();
        if (originalRequestedNodeCount < 1) {
            return 0;
        }
        int cpuCountPerNode = additionalRequired / originalRequestedNodeCount;
        int removableCpuCount = additionalRequired + quotaExceededException.getCurrentUsage() - quotaExceededException.getCurrentLimit();
        int removableNodeCount = (int) Math.ceil((double) removableCpuCount / cpuCountPerNode);

        long provisionableNodeCount = originalRequestedNodeCount - removableNodeCount;
        switch (adjustmentTypeWithThreshold.getAdjustmentType()) {
            case EXACT:
                if (adjustmentTypeWithThreshold.getThreshold() > provisionableNodeCount) {
                    throw new CloudConnectorException(originalRequestedNodeCount + " nodes were requested, but the provisionable node " +
                            "count is " + provisionableNodeCount + ". Threshold is " + adjustmentTypeWithThreshold.getThreshold() +
                            ". Your quota limit exceeded on provider side.", quotaExceededException.getCause());
                }
                break;
            case PERCENTAGE:
                double calculatedPercentage = calculatePercentage(removableNodeCount, originalRequestedNodeCount);
                if (adjustmentTypeWithThreshold.getThreshold() > calculatedPercentage) {
                    throw new CloudConnectorException(originalRequestedNodeCount + " nodes were requested, but the provisionable node " +
                            "count is " + provisionableNodeCount + ". Threshold is " + adjustmentTypeWithThreshold.getThreshold() +
                            "%. Your quota limit exceeded on provider side.", quotaExceededException.getCause());
                }
                break;
            case BEST_EFFORT:
                if (provisionableNodeCount < 1) {
                    throw new CloudConnectorException("We are not able to provision any node. Your quota limit exceeded on provider side.",
                            quotaExceededException.getCause());
                }
                break;
            default:
                throw new CloudConnectorException("Unkown adjustment type: " + adjustmentTypeWithThreshold.getAdjustmentType(),
                        quotaExceededException.getCause());
        }
        return removableNodeCount;
    }

    private double calculatePercentage(double removableNodeCount, double requestedNodeCount) {
        double calculatedPercentage = (requestedNodeCount - removableNodeCount) / requestedNodeCount * ONE_HUNDRED;
        LOGGER.info("Calculated percentage: {}", calculatedPercentage);
        return calculatedPercentage;
    }

    private void decreaseInstances(List<Group> groups, int removableNodeCount) {
        for (Group group : groups) {
            List<CloudInstance> newGroupInstances = group.getInstances().stream()
                    .filter(cloudInstance -> cloudInstance.getInstanceId() == null)
                    .collect(Collectors.toList());
            int removableNodeCountForGroup = Integer.min(newGroupInstances.size(), removableNodeCount);
            int fromIndex = newGroupInstances.size() - removableNodeCountForGroup;
            List<CloudInstance> removableInstances = newGroupInstances.subList(fromIndex, newGroupInstances.size());
            removableNodeCount = removableNodeCount - removableInstances.size();
            group.getInstances().removeAll(removableInstances);
        }
    }

    public void fireImageFallbackFlowMessage(Long stackId, String notificationMessage) {
        flowMessageService.fireEventAndLog(stackId, UPDATE_IN_PROGRESS.name(), STACK_UPSCALE_IMAGE_FALLBACK, notificationMessage);
    }

    public InstanceStoreMetadata getInstanceStorageInfo(AuthenticatedContext ac, String instanceType, CloudConnector connector)
            throws CloudbreakServiceException {
        try {
            return connector.metadata().collectInstanceStorageCount(ac, List.of(instanceType));
        } catch (Exception e) {
            String exceptionMessage = format("Exception in vertical scale flow while trying to fetch metadata for the instance type: %s", instanceType);
            LOGGER.warn(exceptionMessage, e);
            throw new CloudbreakServiceException(exceptionMessage, e);
        }
    }
}

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/stack/upscale/StackUpscaleService.kt
package com.sequenceiq.cloudbreak.core.flow2.stack.upscale

import com.sequenceiq.cloudbreak.api.model.InstanceStatus.CREATED
import com.sequenceiq.cloudbreak.api.model.Status.AVAILABLE
import com.sequenceiq.cloudbreak.api.model.Status.UPDATE_IN_PROGRESS
import java.lang.String.format

import java.util.ArrayList
import java.util.HashMap
import java.util.HashSet

import javax.inject.Inject

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

import com.sequenceiq.cloudbreak.api.model.Status
import com.sequenceiq.cloudbreak.cloud.context.CloudContext
import com.sequenceiq.cloudbreak.cloud.event.instance.CollectMetadataResult
import com.sequenceiq.cloudbreak.cloud.event.resource.UpscaleStackResult
import com.sequenceiq.cloudbreak.cloud.model.CloudInstance
import com.sequenceiq.cloudbreak.cloud.model.CloudResource
import com.sequenceiq.cloudbreak.cloud.model.CloudResourceStatus
import com.sequenceiq.cloudbreak.cloud.model.CloudVmMetaDataStatus
import com.sequenceiq.cloudbreak.cloud.model.Group
import com.sequenceiq.cloudbreak.cloud.model.InstanceStatus
import com.sequenceiq.cloudbreak.common.type.BillingStatus
import com.sequenceiq.cloudbreak.converter.spi.InstanceMetaDataToCloudInstanceConverter
import com.sequenceiq.cloudbreak.converter.spi.ResourceToCloudResourceConverter
import com.sequenceiq.cloudbreak.converter.spi.StackToCloudStackConverter
import com.sequenceiq.cloudbreak.core.flow2.stack.FlowMessageService
import com.sequenceiq.cloudbreak.core.flow2.stack.Msg
import com.sequenceiq.cloudbreak.core.flow2.stack.downscale.StackScalingFlowContext
import com.sequenceiq.cloudbreak.domain.InstanceGroup
import com.sequenceiq.cloudbreak.domain.Resource
import com.sequenceiq.cloudbreak.domain.Stack
import com.sequenceiq.cloudbreak.reactor.api.event.StackFailureEvent
import com.sequenceiq.cloudbreak.repository.InstanceGroupRepository
import com.sequenceiq.cloudbreak.repository.StackUpdater
import com.sequenceiq.cloudbreak.service.cluster.ClusterService
import com.sequenceiq.cloudbreak.service.events.CloudbreakEventService
import com.sequenceiq.cloudbreak.service.hostgroup.HostGroupService
import com.sequenceiq.cloudbreak.service.messages.CloudbreakMessagesService
import com.sequenceiq.cloudbreak.service.stack.InstanceMetadataService
import com.sequenceiq.cloudbreak.service.stack.StackService
import com.sequenceiq.cloudbreak.service.stack.connector.OperationException
import com.sequenceiq.cloudbreak.service.stack.flow.MetadataSetupService
import com.sequenceiq.cloudbreak.service.stack.flow.StackScalingService

@Service
class StackUpscaleService {
    @Inject
    private val stackUpdater: StackUpdater? = null
    @Inject
    private val flowMessageService: FlowMessageService? = null
    @Inject
    private val stackScalingService: StackScalingService? = null
    @Inject
    private val stackService: StackService? = null
    @Inject
    private val metadataConverter: InstanceMetaDataToCloudInstanceConverter? = null
    @Inject
    private val cloudResourceConverter: ResourceToCloudResourceConverter? = null
    @Inject
    private val instanceMetadataService: InstanceMetadataService? = null
    @Inject
    private val instanceGroupRepository: InstanceGroupRepository? = null
    @Inject
    private val metadataSetupService: MetadataSetupService? = null
    @Inject
    private val clusterService: ClusterService? = null
    @Inject
    private val messagesService: CloudbreakMessagesService? = null
    @Inject
    private val eventService: CloudbreakEventService? = null
    @Inject
    private val hostGroupService: HostGroupService? = null
    @Inject
    private val cloudStackConverter: StackToCloudStackConverter? = null


    fun startAddInstances(stack: Stack, scalingAdjustment: Int?) {
        val statusReason = format("Adding %s new instance(s) to the infrastructure.", scalingAdjustment)
        stackUpdater!!.updateStackStatus(stack.id, UPDATE_IN_PROGRESS, statusReason)
        flowMessageService!!.fireEventAndLog(stack.id, Msg.STACK_ADDING_INSTANCES, UPDATE_IN_PROGRESS.name, scalingAdjustment)
    }

    fun finishAddInstances(context: StackScalingFlowContext, payload: UpscaleStackResult): Set<Resource> {
        LOGGER.info("Upscale stack result: {}", payload)
        val results = payload.results
        validateResourceResults(context.cloudContext, payload.errorDetails, results)
        updateNodeCount(context.stack.id, context.cloudStack.groups, results, false)
        val resourceSet = transformResults(results, context.stack)
        if (resourceSet.isEmpty()) {
            throw OperationException("Failed to upscale the cluster since all create request failed: " + results.get(0).statusReason)
        }
        LOGGER.debug("Adding new instances to the stack is DONE")
        return resourceSet
    }

    fun finishExtendMetadata(stack: Stack, instanceGroupName: String, payload: CollectMetadataResult): Set<String> {
        val coreInstanceMetaData = payload.results
        metadataSetupService!!.saveInstanceMetaData(stack, coreInstanceMetaData, CREATED)
        val upscaleCandidateAddresses = HashSet<String>()
        for (cloudVmMetaDataStatus in coreInstanceMetaData) {
            upscaleCandidateAddresses.add(cloudVmMetaDataStatus.metaData.privateIp)
        }
        val instanceGroup = instanceGroupRepository!!.findOneByGroupNameInStack(stack.id, instanceGroupName)
        val nodeCount = instanceGroup.nodeCount!! + coreInstanceMetaData.size
        instanceGroup.nodeCount = nodeCount
        instanceGroupRepository.save(instanceGroup)
        clusterService!!.updateClusterStatusByStackId(stack.id, AVAILABLE)
        eventService!!.fireCloudbreakEvent(stack.id, BillingStatus.BILLING_CHANGED.name,
                messagesService!!.getMessage("stack.metadata.setup.billing.changed"))
        flowMessageService!!.fireEventAndLog(stack.id, Msg.STACK_METADATA_EXTEND, AVAILABLE.name)

        return upscaleCandidateAddresses
    }

    fun finishExtendHostMetadata(stack: Stack) {
        stackUpdater!!.updateStackStatus(stack.id, AVAILABLE, "Stack upscale has been finished successfully.")
        flowMessageService!!.fireEventAndLog(stack.id, Msg.STACK_UPSCALE_FINISHED, AVAILABLE.name)
    }

    fun handleStackUpscaleFailure(stack: Stack, payload: StackFailureEvent) {
        LOGGER.error("Exception during the downscaling of stack", payload.exception)
        try {
            val errorReason = payload.exception.message
            stackUpdater!!.updateStackStatus(stack.id, AVAILABLE, "Stack update failed. " + errorReason)
            flowMessageService!!.fireEventAndLog(stack.id, Msg.STACK_INFRASTRUCTURE_UPDATE_FAILED, AVAILABLE.name, errorReason)
        } catch (e: Exception) {
            LOGGER.error("Exception during the handling of stack scaling failure: {}", e.message)
        }

    }

    private fun transformResults(cloudResourceStatuses: List<CloudResourceStatus>, stack: Stack): Set<Resource> {
        val retSet = HashSet<Resource>()
        for (cloudResourceStatus in cloudResourceStatuses) {
            if (!cloudResourceStatus.isFailed) {
                val cloudResource = cloudResourceStatus.cloudResource
                val resource = Resource(cloudResource.type, cloudResource.name, cloudResource.reference, cloudResource.status,
                        stack, null)
                retSet.add(resource)
            }
        }
        return retSet
    }

    private fun validateResourceResults(cloudContext: CloudContext, exception: Exception?, results: List<CloudResourceStatus>) {
        if (exception != null) {
            LOGGER.error(format("Failed to upscale stack: %s", cloudContext), exception)
            throw OperationException(exception)
        }
        if (results.size == 1 && (results[0].isFailed || results[0].isDeleted)) {
            throw OperationException(format("Failed to upscale the stack for %s due to: %s", cloudContext, results[0].statusReason))
        }
    }

    private fun updateNodeCount(stackId: Long?, originalGroups: List<Group>, statuses: List<CloudResourceStatus>, create: Boolean) {
        for (group in originalGroups) {
            val nodeCount = group.instances.size
            val failedResources = removeFailedMetadata(stackId, statuses, group)
            if (!failedResources.isEmpty() && create) {
                val failedCount = failedResources.size
                val instanceGroup = instanceGroupRepository!!.findOneByGroupNameInStack(stackId, group.name)
                instanceGroup.nodeCount = nodeCount - failedCount
                instanceGroupRepository.save(instanceGroup)
                flowMessageService!!.fireEventAndLog(stackId, Msg.STACK_INFRASTRUCTURE_ROLLBACK_MESSAGE, Status.UPDATE_IN_PROGRESS.name,
                        failedCount, group.name, failedResources[0].statusReason)
            }
        }
    }

    private fun removeFailedMetadata(stackId: Long?, statuses: List<CloudResourceStatus>, group: Group): List<CloudResourceStatus> {
        val failedResources = HashMap<Long, CloudResourceStatus>()
        val groupPrivateIds = getPrivateIds(group)
        for (status in statuses) {
            val privateId = status.privateId
            if (privateId != null && status.isFailed && !failedResources.containsKey(privateId) && groupPrivateIds.contains(privateId)) {
                failedResources.put(privateId, status)
                instanceMetadataService!!.deleteInstanceRequest(stackId, privateId)
            }
        }
        return ArrayList(failedResources.values)
    }

    private fun getPrivateIds(group: Group): Set<Long> {
        val ids = HashSet<Long>()
        for (cloudInstance in group.instances) {
            ids.add(cloudInstance.template!!.privateId)
        }
        return ids
    }

    fun getNewInstances(stack: Stack): List<CloudInstance> {
        val cloudInstances = cloudStackConverter!!.buildInstances(stack)
        val iterator = cloudInstances.iterator()
        while (iterator.hasNext()) {
            if (iterator.next().template!!.status !== InstanceStatus.CREATE_REQUESTED) {
                iterator.remove()
            }
        }
        return cloudInstances
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(StackUpscaleService::class.java)
    }
}
